More backends

iankronquist opened the pull request:
Mostly silly stuff. When I wrote a tutorial on how to write a redis backend this happened. It was easy, so I just chucked it in a commit with another inane backend. Also, test generators are fun. Incidentally fixes 
#3.

iankronquist committed:
Don't do silly things like repeat a value

iankronquist committed:
Add a dictionary and a redis backend

iankronquist committed:
Add an API conformation test

Generators are fun!

iankronquist commented:
@willkg r?

willkg commented on spicedham/dictwrapper.py:
```diff
@@ -0,0 +1,23 @@
+from spicedham.backend import BaseBackend
+
+class DictWrapper(BaseBackend):
```
I don't really understand what this is. Can you add a docstring on the class that explains what this does and why? Seems like it stores things in a class property dict, but I'm not sure why.

willkg commented on spicedham/dictwrapper.py:
```diff
@@ -0,0 +1,23 @@
+        """
+        Classifier and key are strings. value may be anything.
+        """
+        self.store[classifier + key] = value
```
When composing a key like this, you need to be careful to compose them in a way that reduces the likelihood you end up with two combinations that produce the same key. e.g.
- classifier: "foobar" key: "baz"
- classifier: "foo" key: "barbaz"

Those both end up with a key of "foobarbaz".

Some possibilities:
1. hash all the parts and use a delimiter like `::`
2. put them in a tuple (tuples are hashable, so you can use them as dict keys)
3. use a hierarchical store of self.store -> classifier -> key -> value
4. some fourth option?

willkg commented on spicedham/rediswrapper.py:
```diff
@@ -0,0 +1,40 @@
+        """
+        classifier and key are strings which will be concatenated to form a
+        simgle key. value is a jsonifiable dictionary.
+        """
```
The docstrings for set_key, reset, and get_key are the same for all BaseBackends, so you can just document it on the superclass and skip it on all the subclasses.

willkg commented on tests/test_backend_api_conformation.py:
```diff
@@ -0,0 +1,75 @@
+
+try:
+    from spicedham.rediswrapper import RedisWrapper
+except ImportError as e:
```
Nit: You don't use `e`, so you don't need to capture it in a name.

willkg commented on tests/test_backend_api_conformation.py:
```diff
@@ -0,0 +1,75 @@
+        sub_obj = subclass({'engine': 'sqlite:///./db.sqlite'})
+        yield _test_subclass_reset, sub_obj
+        yield _test_subclass_set_and_get, sub_obj
+        yield _test_subclass_set_list_and_get_list, sub_obj
```
This test runs for some arbitrary number of subclasses, but between test runs you don't get any good indication of whether it ran for all the subclasses you wanted it to run for or not.

A better way to do this is to have a test class for each backend and throw a SkipTest (or equivalent) if the test can't run because the dependencies aren't available for the backend. That way when you run the tests, you see successes, failures and the tests that were skipped.

iankronquist committed:
Docstring fixes

iankronquist committed:
Be smarter about keys

iankronquist committed:
Be smarter about hashing keys

iankronquist committed:
Remove unnecessary docstrings

iankronquist committed:
Test individual backends as opposed to all at once

I used to just test API conformance for all of the backends by finding
all of the BaseBackend subclasses and using a tes generator for each of them.
Now I test the backends individually so we can skip them more intelligently and
see more precisely why they failed.

iankronquist committed:
Add a sane default when an engine isn't specified

iankronquist committed:
Test individual backends as opposed to all at once

I used to just test API conformance for all of the backends by finding
all of the BaseBackend subclasses and using a tes generator for each of them.
Now I test the backends individually so we can skip them more intelligently and
see more precisely why they failed

iankronquist commented:
@willkg r? 
I made a silly mistake with commit messages and also changed SqlAlchemyWrapper a bit.

willkg commented on tests/test_backend_api_conformation.py:
```diff
@@ -0,0 +1,133 @@
+            except redis.ConnectionError:
+                raise SkipTest('No Redis server on localhost')
+        except ImportError:
+            raise SkipTest('Redis not installed')
```
Nice!

willkg commented:
Looks good to me! r+
